# Day 1 — Python, NumPy, Pandas and EDA

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/tech4alltraining/aiml/blob/main/mlai-genai-internship/notebooks/01_day1_python_data.ipynb)

**ML/AI & GenAI Internship** · Each concept: examples → 5 exercises → quiz → tasks.

---

## What this notebook covers

Four concepts. **Each one follows the same shape**, so you always know what comes next:

```text
📘 Examples          2-3 worked examples you run and read
✏️ Practice now      5 short exercises, with solutions
🧠 Quick quiz        5 questions to check you understood
🎯 Tasks             2-3 longer pieces for after the session
```

| # | Concept |
|---|---|
| 1 | Python refresher |
| 2 | NumPy: fast arrays |
| 3 | Pandas: tables of data |
| 4 | Exploratory Data Analysis |

**How to use this:** run each cell with `Shift+Enter`. Read the explanation *before* the cell. Never open a solutions block until you have attempted all five exercises.

In [1]:
# Setup. Run this first, every session.

import sys
import numpy as np
import pandas as pd

# Every dataset loads straight from a URL - nothing to download.
BASE = "https://raw.githubusercontent.com/tech4alltraining/aiml/refs/heads/main/datasets/"

print("Python:", sys.version.split()[0])
print("NumPy :", np.__version__)
print("Pandas:", pd.__version__)
print("\nReady.")

Python: 3.12.11
NumPy : 2.4.0
Pandas: 2.3.3

Ready.


---

# Concept 1 — Python refresher

You need six things fluently: **variables, collections, indexing, conditionals, loops, functions.** Nothing more, for now.

🧠 **Analogy: the four containers.** Python's collections are four kinds of container.

| Type | Analogy | Ordered? | Changeable? | Duplicates? |
|---|---|---|---|---|
| **list** `[ ]` | A shelf of books | Yes | Yes | Yes |
| **tuple** `( )` | A sealed box of coordinates | Yes | **No** | Yes |
| **dict** `{k: v}` | A labelled drawer | Yes | Yes | Keys unique |
| **set** `{ }` | A bag of raffle tickets | No | Yes | **No** |

Nine times out of ten the answer is a **list** or a **dict**.

## 📘 Examples

Three worked examples. Run each, then read the output next to the code.

### Example 1 — variables, collections, indexing

In [2]:
# Python works out the type for you; you never declare it.
name = "Asha"        # str
age = 21             # int
score = 87.5         # float
enrolled = True      # bool

print(name, age, score, enrolled)
print(type(name).__name__, type(age).__name__, type(score).__name__)

# The four collections
marks = [78, 92, 65, 88]                 # list  - ordered, changeable
point = (10.5, 20.3)                     # tuple - fixed once made
student = {"name": "Asha", "marks": 87}  # dict  - looked up by key
subjects = {"Math", "Physics", "Math"}   # set   - duplicates vanish

print("\nlist :", marks)
print("tuple:", point)
print("dict :", student)
print("set  :", subjects, " <- only ONE 'Math'")

# Indexing counts from 0. A slice INCLUDES start, EXCLUDES stop.
print("\nmarks[0]   =", marks[0],   " first item")
print("marks[-1]  =", marks[-1],  " last item")
print("marks[1:3] =", marks[1:3], " items 1 and 2, NOT 3")

Asha 21 87.5 True
str int float

list : [78, 92, 65, 88]
tuple: (10.5, 20.3)
dict : {'name': 'Asha', 'marks': 87}
set  : {'Math', 'Physics'}  <- only ONE 'Math'

marks[0]   = 78  first item
marks[-1]  = 88  last item
marks[1:3] = [92, 65]  items 1 and 2, NOT 3


### Example 2 — conditionals and loops

In [3]:
average = sum(marks) / len(marks)

# elif means "otherwise, if". Using a second `if` is a classic bug:
# both branches could then run and the later one wins.
if average >= 90:
    grade = "A"
elif average >= 75:
    grade = "B"
else:
    grade = "C"

print(f"Average {average:.2f} -> grade {grade}")

# enumerate() gives position AND value, so you never need a counter variable.
print()
for index, mark in enumerate(marks):
    status = "pass" if mark >= 75 else "needs work"
    print(f"  Subject {index + 1}: {mark:>3}  ({status})")

Average 80.75 -> grade B

  Subject 1:  78  (pass)
  Subject 2:  92  (pass)
  Subject 3:  65  (needs work)
  Subject 4:  88  (pass)


### Example 3 — functions and list comprehensions

In [4]:
def calculate_percentage(obtained, total=100):
    """Return the percentage, rounded to 2 places. total defaults to 100."""
    return round((obtained / total) * 100, 2)

print(calculate_percentage(87))             # uses the default
print(calculate_percentage(87, total=120))  # overrides it

# List comprehension: build a list in one line.
# You will READ this constantly in ML code, so learn to recognise it.
squared = [m ** 2 for m in marks]           # transform every item
passed  = [m for m in marks if m >= 75]     # keep only some items

print("\noriginal:", marks)
print("squared :", squared)
print("passed  :", passed)

# The same thing written the long way, for comparison:
passed_long = []
for m in marks:
    if m >= 75:
        passed_long.append(m)
print("same    :", passed_long)

87.0
72.5

original: [78, 92, 65, 88]
squared : [6084, 8464, 4225, 7744]
passed  : [78, 92, 88]
same    : [78, 92, 88]


## ✏️ Practice now — Python refresher

Five short exercises. Do them **before** moving on — this is where the learning happens.

**1.** From `nums = [12, 45, 7, 89, 23, 56]`, print the **first**, the **last**, and the **middle two** items.

**2.** Write a function `is_even(n)` that returns `True` or `False`. Test it on 4 and 7.

**3.** Using a **list comprehension**, build a list of only the numbers in `nums` that are above 30.

**4.** Build a dict `{'Asha': 87, 'Ravi': 74, 'Meera': 95}` and print the name of the student with the **highest** mark.

**5.** Loop over `nums` and print `"even"` or `"odd"` next to each number.

In [5]:
nums = [12, 45, 7, 89, 23, 56]

# 1.


# 2.


# 3.


# 4.


# 5.

<details>
<summary><b>Solutions — open only after you have tried all five</b></summary>

```python
nums = [12, 45, 7, 89, 23, 56]

# 1
print(nums[0], nums[-1], nums[2:4])

# 2
def is_even(n):
    return n % 2 == 0

print(is_even(4), is_even(7))

# 3
print([n for n in nums if n > 30])

# 4
scores = {"Asha": 87, "Ravi": 74, "Meera": 95}
print(max(scores, key=scores.get))       # max() with key= compares by VALUE

# 5
for n in nums:
    print(n, "even" if n % 2 == 0 else "odd")
```

</details>

## 🧠 Quick quiz — Python refresher

Answer all five from memory. No scrolling back.

**Q1.** `marks = [78, 92, 65, 88]`. What does `marks[1:3]` return, and how many items is that?

**Q2.** What is the difference between a list and a tuple?

**Q3.** Why does `{'Math', 'Physics', 'Math'}` print only two items?

**Q4.** In `def f(x, total=100):`, what is `total=100` called and what does it do?

**Q5.** Rewrite `[m for m in marks if m >= 75]` as an ordinary `for` loop.

<details>
<summary><b>Answers</b></summary>

**A1.** `[92, 65]` — two items. A slice **includes** the start index and **excludes** the stop index.

**A2.** A list can be changed after it is created; a tuple cannot. Use a tuple for things that should stay fixed, like a pair of coordinates.

**A3.** It is a **set**, and sets silently refuse duplicates. There is no error — the second `'Math'` is simply dropped.

**A4.** A **default argument**. Callers may leave it out, and it then takes the value 100. Omit a *required* argument and you get a `TypeError`.

**A5.** ```python\nresult = []\nfor m in marks:\n    if m >= 75:\n        result.append(m)\n```\nThe comprehension is the same three lines compressed into one.

</details>

## 🎯 Tasks — Python refresher

Longer work. Do these after the session, in your own time.

### Task 1: Fix the broken program

This program should print each student's average and grade. It has **four** bugs.

```python
students = {"Asha": [78, 92, 65], "Ravi": [88, 74, 91], "Meera": [95, 89, 97]}

def average(marks)
    return sum(marks) / len(mark)

for name in students:
    avg = average(students[name])
    if avg > 90:
        grade = "A"
    if avg > 75:
        grade = "B"
    else:
        grade = "C"
    print(name + " scored " + avg + " grade " + grade)
```

Find and fix all four. **Read each error's last line first** — it names the actual problem.

<details><summary>Hints</summary>

1. Line 3 is missing a character every `def` line needs.
2. Line 4 uses a variable name that does not exist — look very closely.
3. The second `if` should be something else, or Meera gets the wrong grade.
4. You cannot join a number to a string with `+`. Use an f-string.
</details>

### Task 2: Marks report generator

Write a program that, given a dict of `{name: [marks]}`:

- prints each student's average to 1 decimal place
- prints their grade (A ≥ 90, B ≥ 75, C ≥ 60, else F)
- prints the class average
- prints the name of the top student and the bottom student
- counts how many students passed (average ≥ 60)

Use at least one function and one list comprehension.

### Task 3: Word counter

Given a paragraph of text, print:

- the total number of words
- the number of **unique** words (hint: a set)
- the five longest words
- how many words are longer than 6 letters

Test it on a paragraph from this notebook.

---

# Concept 2 — NumPy: fast arrays

🧠 **Analogy: the shopping bag and the egg tray.**

A Python **list** is a shopping bag — throw anything in, but to act on everything you must pull items out one at a time.

A NumPy **array** is an egg tray — every slot holds the same kind of thing, in a fixed grid. Because the computer knows every slot is identical, it can act on **all of them at once**.

On a million numbers the egg tray is roughly fifty times faster. **That is the entire reason NumPy exists.**

## 📘 Examples

### Example 1 — the difference that matters

In [6]:
python_list = [100, 200, 300]
numpy_array = np.array([100, 200, 300])

print("list  * 2 =", python_list * 2)
print("            ^ the BAG was duplicated - probably not what you wanted")
print()
print("array * 2 =", numpy_array * 2)
print("            ^ every ELEMENT was doubled - this is what we want")

list  * 2 = [100, 200, 300, 100, 200, 300]
            ^ the BAG was duplicated - probably not what you wanted

array * 2 = [200 400 600]
            ^ every ELEMENT was doubled - this is what we want


### Example 2 — shapes, and maths on whole arrays

In [7]:
a = np.array([1, 2, 3, 4, 5])
b = np.array([[1, 2, 3],
              [4, 5, 6]])

print("a       :", a)
print("a.shape :", a.shape, "  5 elements, 1 dimension")
print("b.shape :", b.shape, " 2 rows, 3 columns")
print("a.dtype :", a.dtype, "  every element is this type")

# Every operation applies to the WHOLE array. No loop anywhere.
print("\na + 10  :", a + 10)
print("a ** 2  :", a ** 2)
print("a / 2   :", a / 2)

a       : [1 2 3 4 5]
a.shape : (5,)   5 elements, 1 dimension
b.shape : (2, 3)  2 rows, 3 columns
a.dtype : int64   every element is this type

a + 10  : [11 12 13 14 15]
a ** 2  : [ 1  4  9 16 25]
a / 2   : [0.5 1.  1.5 2.  2.5]


### Example 3 — statistics and boolean masking

In [8]:
data = np.array([23, 45, 12, 67, 34, 89, 21])

print("data  :", data)
print("mean  :", data.mean().round(2))
print("median:", np.median(data))
print("std   :", data.std().round(2))
print("min/max:", data.min(), data.max())

# BOOLEAN MASKING - the idea behind every Pandas filter you will ever write.
mask = data > 30                 # an array of True/False, one per element

print("\ndata > 30  :", mask)
print("data[mask] :", data[mask], " <- only the True positions")
print("how many   :", mask.sum(), "  (True counts as 1)")

data  : [23 45 12 67 34 89 21]
mean  : 41.57
median: 34.0
std   : 25.66
min/max: 12 89

data > 30  : [False  True False  True  True  True False]
data[mask] : [45 67 34 89]  <- only the True positions
how many   : 4   (True counts as 1)


## ✏️ Practice now — NumPy

Five short exercises. Do them **before** moving on — this is where the learning happens.

**1.** Create `marks = np.array([78, 92, 65, 88, 45, 97, 55, 71, 83, 60])` and print its **mean** to 1 decimal place.

**2.** Print the **highest** and **lowest** mark.

**3.** Using a boolean mask, count how many students scored **75 or above**.

**4.** Print only the marks **below 60**.

**5.** Add 5 to every mark but **cap the result at 100** (hint: `np.minimum`).

In [9]:
marks = np.array([78, 92, 65, 88, 45, 97, 55, 71, 83, 60])

# 1.


# 2.


# 3.


# 4.


# 5.

<details>
<summary><b>Solutions — open only after you have tried all five</b></summary>

```python
marks = np.array([78, 92, 65, 88, 45, 97, 55, 71, 83, 60])

# 1
print("mean:", round(marks.mean(), 1))

# 2
print("max/min:", marks.max(), marks.min())

# 3
print("scored 75+:", (marks >= 75).sum())

# 4
print("below 60:", marks[marks < 60])

# 5
print("boosted:", np.minimum(marks + 5, 100))
```

</details>

## 🧠 Quick quiz — NumPy

Answer all five from memory. No scrolling back.

**Q1.** `[1, 2, 3] * 2` and `np.array([1, 2, 3]) * 2` give different results. What are they, and why?

**Q2.** What does `.shape` of `(2, 3)` tell you?

**Q3.** What does `(data > 30).sum()` count, and why does `.sum()` work on True/False?

**Q4.** Why do almost all ML examples set a random seed, e.g. `default_rng(seed=42)`?

**Q5.** Why does this course tell you to avoid `for` loops over arrays?

<details>
<summary><b>Answers</b></summary>

**A1.** The list gives `[1, 2, 3, 1, 2, 3]` — it repeats the container. The array gives `[2, 4, 6]` — it doubles each element. A list is a bag; an array is an egg tray where every slot is the same type, so maths applies element-wise.

**A2.** Two rows and three columns — a 2-D array with 6 elements in total.

**A3.** It counts how many elements are above 30. `True` counts as 1 and `False` as 0, so summing a boolean array counts the `True`s.

**A4.** So the 'random' numbers are the same every run. You and your classmate then get identical results, and your work is reproducible when a marker re-runs it.

**A5.** Because NumPy applies operations to the whole array at once, which is far faster — and in real ML you work with a million rows, not ten. Thinking in whole arrays is the habit you need.

</details>

## 🎯 Tasks — NumPy

Longer work. Do these after the session, in your own time.

### Task 1: Class statistics report

Given a 2-D array where each **row** is a student and each **column** is a subject:

```python
scores = np.array([[78, 92, 65],
                   [88, 74, 91],
                   [95, 89, 97],
                   [45, 60, 55]])
```

Without any `for` loop, print:
- each student's average (hint: `axis=1`)
- each subject's average (hint: `axis=0`)
- the highest single score, and which student and subject it belongs to
- how many scores in the whole table are below 60

### Task 2: Speed comparison

Prove the speed claim yourself.

Create a Python list and a NumPy array, each with one million numbers. Square every element, timing both. Use `import time` or the `%%timeit` magic.

Report the ratio. Was it close to 50×?

---

# Concept 3 — Pandas: tables of data

🧠 **Analogy: a spreadsheet that takes orders.**

A `DataFrame` is a spreadsheet — rows, columns, headers. The difference is that instead of clicking and dragging, you give it instructions.

| What you would do in Excel | What you type in Pandas |
|---|---|
| Scroll to the top to check the data | `df.head()` |
| Read the row count off the status bar | `df.shape` |
| Filter → "Price greater than 100000" | `df[df["price"] > 100000]` |
| Sort descending by Price | `df.sort_values("price", ascending=False)` |
| Insert a PivotTable | `df.groupby("fuel")["price"].mean()` |
| Find blanks with conditional formatting | `df.isnull().sum()` |
| Add a formula column | `df["new"] = df["a"] / df["b"]` |

**You are not learning a new idea — you are learning new words for one you already have.**

## 📘 Examples

### Example 1 — load a dataset and look at it

In [10]:
# 15,000 used cars with their selling prices.
df = pd.read_csv(BASE + "regression/cardekho_dataset.csv")

print("Rows and columns:", df.shape)
df.head()

Rows and columns: (15411, 14)


,Unnamed: 0,car_name,brand,model,vehicle_age,km_driven,seller_type,fuel_type,transmission_type,mileage,engine,max_power,seats,selling_price
0,0,Maruti Alto,Maruti,Alto,9,120000,Individual,Petrol,Manual,19.70,796,46.30,5,120000
1,1,Hyundai Grand,Hyundai,Grand,5,20000,Individual,Petrol,Manual,18.90,1197,82.00,5,550000
2,2,Hyundai i20,Hyundai,i20,11,60000,Individual,Petrol,Manual,17.00,1197,80.00,5,215000
3,3,Maruti Alto,Maruti,Alto,9,37000,Individual,Petrol,Manual,20.92,998,67.10,5,226000
4,4,Ford Ecosport,Ford,Ecosport,6,30000,Dealer,Diesel,Manual,22.77,1498,98.59,5,570000


### Example 2 — selecting and filtering

In [11]:
# ONE name in [] gives a Series. A LIST of names gives a DataFrame.
print("one column ->", type(df["selling_price"]).__name__)
print("two columns ->", type(df[["brand", "selling_price"]]).__name__)

# FILTERING: use & for AND and | for OR, and wrap EACH condition in brackets.
# Plain `and` / `or` do not work on whole columns.
expensive = df[df["selling_price"] > 1000000]
print("\nCars above 10 lakh:", len(expensive))

recent_petrol = df[(df["vehicle_age"] < 5) & (df["fuel_type"] == "Petrol")]
print("Recent petrol cars:", len(recent_petrol))

recent_petrol[["car_name", "vehicle_age", "selling_price"]].head()

one column -> Series
two columns -> DataFrame

Cars above 10 lakh: 2510
Recent petrol cars: 3098


,car_name,vehicle_age,selling_price
7,Maruti Wagon R,3,410000
8,Hyundai Venue,2,1050000
9,Maruti Swift,4,511000
12,Mini Cooper,4,3250000
28,Hyundai i20,3,700000


### Example 3 — grouping, sorting and new columns

In [12]:
# GROUPING is the single most useful Pandas skill.
# "For each fuel type, what is the average price?"
print("Average price by fuel type:")
print(df.groupby("fuel_type")["selling_price"].mean().round(0).sort_values(ascending=False))

# A new column from existing ones.
df["price_lakhs"] = (df["selling_price"] / 100000).round(2)

# Sorting.
print("\nThree most expensive cars:")
print(df.sort_values("selling_price", ascending=False)
        [["car_name", "price_lakhs", "fuel_type"]].head(3))

# Counting categories.
print("\nHow many of each fuel type:")
print(df["fuel_type"].value_counts())

Average price by fuel type:
fuel_type
Electric    1853500.0
Diesel      1000469.0
Petrol       572862.0
CNG          417688.0
LPG          206273.0
Name: selling_price, dtype: float64

Three most expensive cars:
                  car_name  price_lakhs fuel_type
3799     Ferrari GTC4Lusso        395.0    Petrol
10969    Rolls-Royce Ghost        242.0    Petrol
1172   Bentley Continental        145.0    Petrol

How many of each fuel type:
fuel_type
Petrol      7643
Diesel      7419
CNG          301
LPG           44
Electric       4
Name: count, dtype: int64


## ✏️ Practice now — Pandas

Five short exercises. Do them **before** moving on — this is where the learning happens.

**1.** Load the Titanic dataset: `titanic = pd.read_csv(BASE + "classification/archive/titanic.csv")`. Print its **shape**.

**2.** Print the **first 3 rows**, then the **column names**.

**3.** How many passengers **survived**? (The `survived` column is 1 or 0, so `.sum()` works.)

**4.** What is the **survival rate by `sex`**? Use `groupby`.

**5.** Filter to passengers in `class == 'First'` who **survived**, and print how many there were.

In [13]:
titanic = pd.read_csv(BASE + "classification/archive/titanic.csv")

# 1.


# 2.


# 3.


# 4.


# 5.

<details>
<summary><b>Solutions — open only after you have tried all five</b></summary>

```python
titanic = pd.read_csv(BASE + "classification/archive/titanic.csv")

# 1
print(titanic.shape)

# 2
print(titanic.head(3))
print(titanic.columns.tolist())

# 3
print("survived:", titanic["survived"].sum())

# 4
print(titanic.groupby("sex")["survived"].mean().round(3))

# 5
first_survivors = titanic[(titanic["class"] == "First") & (titanic["survived"] == 1)]
print("first-class survivors:", len(first_survivors))
```

</details>

## 🧠 Quick quiz — Pandas

Answer all five from memory. No scrolling back.

**Q1.** What is the difference between `df["price"]` and `df[["price"]]`?

**Q2.** Why must you write `df[(a > 1) & (b < 2)]` instead of `df[a > 1 and b < 2]`?

**Q3.** This code prints missing values that you thought you removed. Why?\n```python\ndf.dropna()\nprint(df.isnull().sum())\n```

**Q4.** What does `df.groupby("fuel_type")["price"].mean()` do, in plain English?

**Q5.** Which five commands should you run first on any new dataset, and what does each tell you?

<details>
<summary><b>Answers</b></summary>

**A1.** `df["price"]` returns a **Series** (a single column). `df[["price"]]` returns a **DataFrame** (a table that happens to have one column). scikit-learn wants a DataFrame for `X`, which is why you see double brackets there.

**A2.** `and` works on single True/False values, but you are comparing whole columns. `&` compares element by element. The brackets are required because `&` binds more tightly than `>`.

**A3.** `dropna()` returns a **new** DataFrame and does not modify `df`. You need `df = df.dropna()`. Most Pandas methods return a copy — this catches almost everybody once.

**A4.** Split the rows into groups by fuel type, then for each group compute the average price. It is the Pandas version of a spreadsheet PivotTable.

**A5.** `df.shape` (how big), `df.head()` (do the values look sensible), `df.info()` (column types and non-null counts), `df.describe()` (ranges and outliers), `df.isnull().sum()` (what is missing).

</details>

## 🎯 Tasks — Pandas

Longer work. Do these after the session, in your own time.

### Task 1: The Titanic investigation

Using the Titanic dataset, answer each of these **in one sentence of plain English**, backed by code:

1. What fraction of all passengers survived?
2. Survival rate by `sex` — and by how much do they differ?
3. Survival rate by `class`?
4. What is the average `age`, and how many rows are missing it?
5. Combine sex and class: which single group had the best survival rate, and which the worst?
6. Write one sentence describing who was most likely to survive.

### Task 2: Build your own describe()

Write `my_describe(df)` that returns count, mean, std, min, 25%, 50%, 75% and max for every numeric column — **without** calling `.describe()`.

Compare your output against the real thing on the cardekho dataset.

<details><summary>Hints</summary>

- `df.select_dtypes(include="number").columns` gives the numeric columns
- `series.quantile(0.25)` gives a percentile
- Build a dict of dicts and hand it to `pd.DataFrame`
</details>

### Task 3: Car market report

Using the cardekho dataset, produce a short report answering:

- Which brand has the highest average price, among brands with at least 50 cars? (Why does that minimum matter?)
- How does average price change with `vehicle_age`? Group into age bands.
- Which `seller_type` sells the most expensive cars on average?
- Is there any brand where diesel cars sell for *less* than petrol?

Each answer needs a number **and** a sentence.

---

# Concept 4 — Exploratory Data Analysis

🧠 **Analogy: the doctor's check-up.**

A doctor does not prescribe medicine the moment you walk in. First comes the check-up: temperature, blood pressure, a few questions. Only then a treatment.

**EDA is the check-up for your dataset.** Students who skip it and jump straight to `model.fit()` are prescribing medicine without examining the patient — and then wondering why it did not work.

### The five questions, in this order

1. **How big is it?** `df.shape`
2. **What is each column?** `df.info()`
3. **What is missing?** `df.isnull().sum()`
4. **What are the ranges and outliers?** `df.describe()`
5. **What is the target, and is it balanced?** `df["target"].value_counts()`

## 📘 Examples

### Example 1 — twelve rows you can check with your eyes

In [14]:
# Before working with 15,000 rows, work with one you can hold in your head.
# This file has TWELVE rows. Read it, and spot the problems yourself.

small = pd.read_csv(BASE + "prepreprocessing/pre_data.csv")
small

,Country,Age,Salary,Purchased
0,France,44.0,72000.0,No
1,Spain,27.0,48000.0,Yes
2,NaN,30.0,54000.0,No
3,Spain,38.0,61000.0,No
4,Germany,40.0,NaN,Yes
5,France,35.0,58000.0,Yes
6,Spain,NaN,52000.0,No
7,France,48.0,79000.0,Yes
8,Germany,50.0,83000.0,No
9,France,37.0,67000.0,Yes


<details>
<summary><b>What is wrong with this data — open after you have looked</b></summary>

1. **Row 2 has a missing Country** — the line begins with a comma
2. **Row 4 has a missing Salary**
3. **Row 6 has a missing Age**
4. **Rows 8 and 10 are identical** — Germany, 50, 83000, No
5. **Row 11 has a salary of 1,500,000** — over 18× the next highest

Three out of five is a good first attempt.
</details>

### Example 2 — confirming what you saw, with code

In [15]:
print("--- Missing values per column ---")
print(small.isnull().sum())

print("\n--- Duplicate rows ---")
print(small[small.duplicated(keep=False)])

print("\n--- Numeric summary ---")
print(small.describe().round(1))

--- Missing values per column ---
Country      1
Age          1
Salary       1
Purchased    0
dtype: int64

--- Duplicate rows ---
    Country   Age   Salary Purchased
8   Germany  50.0  83000.0        No
10  Germany  50.0  83000.0        No

--- Numeric summary ---
        Age     Salary
count  11.0       11.0
mean   40.8   196090.9
std     8.2   432635.3
min    27.0    48000.0
25%    36.0    56000.0
50%    40.0    67000.0
75%    49.0    81000.0
max    50.0  1500000.0


### 📊 How to read a `describe()` table

| Row | What it tells you | The check to make |
|---|---|---|
| `count` | How many **non-missing** values | Lower than the row count? You have gaps in this column |
| `mean` | The average | Compare with `50%` — a big gap means outliers are dragging it |
| `std` | How spread out the values are | Larger than the mean? Very wide spread, probably outliers |
| `min` | Smallest value | Negative where it should not be? A data error |
| `25%` / `50%` / `75%` | The quartiles. `50%` is the **median** | These are where the bulk of your data actually lives |
| `max` | Largest value | **Many times bigger than `75%`? That is an outlier** |

**Applied to the Salary column above:**

- `75%` is **81,000** but `max` is **1,500,000** → the outlier, now measured rather than eyeballed
- `mean` is about **196,091** but the median is **67,000** → one row dragged the average up by nearly 130,000

> **That second point is exactly why you fill missing numbers with the median, not the mean.** You have now seen it happen rather than been told it.

### Example 3 — the full five-question sweep on a real dataset

In [16]:
loans = pd.read_csv(BASE + "loan_data_10k.csv")

print("=== 1. How big is it? ===")
print(loans.shape)

print("\n=== 3. What is missing? ===")
missing = loans.isnull().sum()
print(missing[missing > 0])

print("\n=== 5. Is the target balanced? ===")
print(loans["loan_status"].value_counts())
print(loans["loan_status"].value_counts(normalize=True).round(3))

=== 1. How big is it? ===
(10000, 14)

=== 3. What is missing? ===
person_gender          1
loan_int_rate          1
loan_percent_income    1
dtype: int64

=== 5. Is the target balanced? ===
loan_status
1    5000
0    5000
Name: count, dtype: int64
loan_status
1    0.5
0    0.5
Name: proportion, dtype: float64


In [17]:
# Which numeric columns relate to the target?
# Values near +1 or -1 move WITH the target. Near 0 means no LINEAR
# relationship on its own.

loans.corr(numeric_only=True)["loan_status"].sort_values().round(3)

person_income                -0.210
person_age                   -0.039
person_emp_exp               -0.039
cb_person_cred_hist_length   -0.033
credit_score                 -0.002
loan_amnt                     0.042
loan_int_rate                 0.363
loan_percent_income           0.405
loan_status                   1.000
Name: loan_status, dtype: float64

### 📊 How to read a correlation column

| Value you see | What it means |
|---|---|
| **+0.7 to +1.0** | Strong positive — as this column rises, the target rises |
| **+0.3 to +0.7** | Moderate positive |
| **−0.3 to +0.3** | Weak *linear* relationship. **Not the same as "useless"** — the relationship may be curved, or only matter alongside another column |
| **−0.7 to −0.3** | Moderate negative |
| **−1.0 to −0.7** | Strong negative — as this column rises, the target falls |

**Two warnings students routinely trip over:**

1. **A correlation of −0.3 is not "weak" in real data.** In messy human data it is worth investigating. Do not discard a column because it scored 0.2.
2. **Correlation is not causation.** Ice cream sales correlate with drowning deaths. Neither causes the other — hot weather causes both.

> **In one sentence:** this column tells you which features are *worth a closer look*, not which features to keep.

## ✏️ Practice now — EDA

Five short exercises. Do them **before** moving on — this is where the learning happens.

**1.** Load the diabetes dataset: `dia = pd.read_csv(BASE + "classification/diabetes_prediction_dataset.csv")`. Print its shape.

**2.** How many missing values are there, and in which columns?

**3.** Is the target (`diabetes`) balanced? Print the proportions and say whether accuracy alone would be a safe metric.

**4.** What is the **average `bmi`** for people with diabetes versus without? Use `groupby`.

**5.** Which numeric column correlates most strongly with `diabetes`?

In [18]:
dia = pd.read_csv(BASE + "classification/diabetes_prediction_dataset.csv")

# 1.


# 2.


# 3.


# 4.


# 5.

<details>
<summary><b>Solutions — open only after you have tried all five</b></summary>

```python
dia = pd.read_csv(BASE + "classification/diabetes_prediction_dataset.csv")

# 1
print(dia.shape)

# 2
missing = dia.isnull().sum()
print(missing[missing > 0] if missing.sum() else "No missing values")

# 3
print(dia["diabetes"].value_counts(normalize=True).round(3))
# Only ~8.5% positive -> heavily imbalanced -> accuracy alone would mislead.

# 4
print(dia.groupby("diabetes")["bmi"].mean().round(2))

# 5
print(dia.corr(numeric_only=True)["diabetes"].drop("diabetes").abs().sort_values(ascending=False).head(3))
```

</details>

## 🧠 Quick quiz — EDA

Answer all five from memory. No scrolling back.

**Q1.** Why do EDA at all, instead of going straight to `model.fit()`?

**Q2.** In a `describe()` output, `mean` is 196,091 and `50%` is 67,000. What does that gap tell you?

**Q3.** Your target is 92% class 0 and 8% class 1. What does that mean for your choice of metric?

**Q4.** Two columns correlate at 0.85. Does one cause the other?

**Q5.** A column correlates with the target at 0.15. Should you delete it?

<details>
<summary><b>Answers</b></summary>

**A1.** Because a model cannot tell you that your data is broken — it will happily train on duplicated rows, mis-scaled units and leaked columns, and give you a confident wrong answer. EDA is the check-up before the prescription.

**A2.** The mean is far above the median, so a small number of very large values are dragging the average up. There are outliers in this column, and the median is the more honest summary.

**A3.** The classes are imbalanced, so accuracy is misleading — always predicting class 0 scores 92% and catches nothing. Report recall, precision and F1 for the minority class instead.

**A4.** No. Correlation measures that they move together, not why. A third factor may drive both, or the link may be coincidence. Causation needs an experiment or domain knowledge, not a correlation number.

**A5.** Not on that basis. Correlation only measures **linear** relationship on its own. The column may matter in combination with another, or its relationship may be curved. Test it in a model before discarding it.

</details>

## 🎯 Tasks — EDA

Longer work. Do these after the session, in your own time.

### Task 1: The data quality report

Pick a dataset from [`datasets/`](../../datasets/) that was **not** used in class. Produce a one-page report covering:

1. Shape, and what one row represents
2. Every column: its type, and whether it is a feature or the target
3. Missing values: which columns, how many, and your recommended fix for **each**
4. Outliers: at least one column checked properly, with your judgement on whether to keep them
5. Duplicates: how many, and whether they are genuine repeats or data errors
6. The target: balanced or imbalanced, and what that means for your metric choice
7. Three questions this dataset could answer
8. **One thing you found that surprised you**

### Task 2: Compare two datasets

Run the five-question sweep on **both** the loan dataset and the diabetes dataset.

Produce a side-by-side table comparing: rows, columns, missing values, target balance, and the strongest correlated feature.

Then write three sentences on which dataset you would rather model, and why.

### Task 3: The suspicious spreadsheet

> *A colleague sends `sales_data.csv` and says "it's clean, just build the model". You find: 3% of `revenue` values are negative, `region` contains both `"North"` and `"north"`, 12% of `customer_id` values are duplicated, and `date` is text in three different formats.*

Write a short memo (under 400 words):

1. Which problems you can fix yourself, and how
2. Which need your colleague's input, and the **exact questions** you would ask
3. What you would do if they replied "just drop the bad rows"
4. Whether you would proceed, and why

**What this really tests:** whether you push back. Dropping the bad rows discards 12% of the data and may bias everything downstream. Negative revenue might be refunds — genuine and meaningful — or an error. **You cannot tell from the data alone.**

---
# ✅ Day 1 exit task

Using the **cardekho** dataset, answer these. Write each as **one sentence in plain English**, not a raw code output.

1. How many cars, and how many columns describe each one?
2. Which fuel type has the highest average selling price?
3. Which brand appears most often?
4. Are there missing values? In which columns?
5. What is the most expensive car, and is that price plausible or a data-entry error?

In [19]:
# Your exit task working here.

# 1.
print("1.", df.shape)

# 2.

# 3.

# 4.

# 5.

1. (15411, 15)


---
## What next

| | |
|---|---|
| **Next notebook** | [Day 2 — Visualisation and supervised learning](02_day2_ml_basics.ipynb) |
| **Handbook** | [Day 1 chapter](../student-handbook.md#day-1--python-refresher-and-data-handling) |
| **More practice** | [exercises-assignments.md](../exercises-assignments.md) |
| **Stuck?** | [Troubleshooting](../troubleshooting.md) |